### PR TITLE
## [v2.0.1] - 2024-01-22

### PR DESCRIPTION
**Pull request by Warhammer4000**
### Added

+ Added Cloud Save Core Package as Dependency
+ Added FileProcessor.cs , replacing the deleted SaveProcessor.cs.
+ Added SaveSlotProcessor.cs.
+ Added new samples for demonstration and their configurations.
+ Added Archiver directory and ArchiverBase.cs, ZipUtilityArchiver.cs.
+ Added Encryptor directory and EncryptorBase.cs, AESEncryptor.cs , and deleted Stitcher.cs.
+ Added ItemDataBehaviour.cs and PlayerDataBehaviour.cs scripts for demo.

## Summary by CodeRabbit

- **New Features**
  - Introduced Cloud Save functionality as a core dependency.
  - Added new components for data processing and encryption.
  - Implemented UnityEvents for Save and Load operations completion.
  - Provided new demonstration samples and configurations.
  - Enhanced Save System with additional methods and properties for improved data management.
  - Created a default cloud provider asset for SaveSystem.

- **Improvements**
  - Extended SaveSystem editor with additional GUI elements and functionalities.
  - Improved SaveSlot data structure with a dictionary for file keys.
  - Added IsDirty property to track modified data states.

- **Documentation**
  - Updated CHANGELOG.md to reflect new version details and feature additions.

- **Bug Fixes**
  - Fixed issues related to save slot metadata retrieval and processing.

- **Refactor**
  - Refactored SaveSystem internals to optimize performance and maintainability.
  - Renamed certain fields and methods for clarity and consistency.

### CODE REVIEW

File: Assets/Packages/com.studio23.ss2.savesystem/Runtime/Core/SaveSystem.cs
Comment (N-Evan) [inline Assets/Packages/com.studio23.ss2.savesystem/Runtime/Core/SaveSystem.cs:1]: @coderabbitai generate unit tests for this file.

File: Assets/Packages/com.studio23.ss2.savesystem/Runtime/Core/SaveSlotProcessor.cs [hunk 1]
```diff
@@ -16,23 +19,34 @@ namespace Studio23.SS2.SaveSystem.Core
...
         [SerializeField] internal SlotConfiguration _slotConfiguration;
 
-        private async void Start()
+        public async UniTask Initialize()
         {
             _fileProcessor = GetComponent<FileProcessor>();
             CreateSlotFolders();
+            await _cloudSaveManager.Initialize();
+            await GetAllCloudMeta();
             await SelectSlot(0);
```
Comment (N-Evan) [inline Assets/Packages/com.studio23.ss2.savesystem/Runtime/Core/SaveSlotProcessor.cs]: Why are we defaulting to `Slot 0` on initialize? Should be assigned when we actually select a slot during our first playthrough of the game and later be loaded depending on the save slot that was used last.

File: Assets/Packages/com.studio23.ss2.savesystem/Runtime/Core/SaveSlotProcessor.cs [hunk 2]
```diff
@@ -57,11 +71,11 @@ internal void CreateSlotFolders()
...
         }
 
-        internal async UniTask<SaveSlot> GetSaveSlotMetaData(int index)
+        internal async UniTask<SaveSlot> GetSaveSlotMetaDataLocal(int index)
         {
-            SaveSlot slotMeta=new SaveSlot($"Save_Slot_{index}");
+            SaveSlot slotMeta=new SaveSlot(index);
 
-            string slotMetaPath = Path.Combine(_slotConfiguration.SavePathRoot, $"Save_Slot_{index}", _slotConfiguration.SlotMetafileName);
+            string slotMetaPath = Path.Combine(_slotConfiguration.SavePathRoot, slotMeta.Name, _slotConfiguration.SlotMetafileName);
```
Comment (BDeshiDev) [inline Assets/Packages/com.studio23.ss2.savesystem/Runtime/Core/SaveSlotProcessor.cs:76]: should be method on saveslot

File: Assets/Packages/com.studio23.ss2.savesystem/Runtime/Core/SaveSlotProcessor.cs [hunk 3]
```diff
@@ -71,14 +85,45 @@ internal async UniTask<SaveSlot> GetSaveSlotMetaData(int index)
...
         }
 
+        internal async UniTask GetAllCloudMeta()
+        {
+            List<UniTask> cloudMetaDownloadTaskList = new List<UniTask>();
+
+            for(int i=0;i< _slotConfiguration.SlotCount; i++)
+            {
+                SaveSlot slotMeta = new SaveSlot(i);
+                string slotMetaPath = Path.Combine(_slotConfiguration.SavePathRoot, slotMeta.Name, _slotConfiguration.SlotMetafileName);
```
Comment (BDeshiDev) [inline Assets/Packages/com.studio23.ss2.savesystem/Runtime/Core/SaveSlotProcessor.cs:93]: Repeated a lot. should be method on saveslot.

File: Assets/Packages/com.studio23.ss2.savesystem/Runtime/Core/SaveSlotProcessor.cs [hunk 4]
```diff
@@ -71,14 +85,45 @@ internal async UniTask<SaveSlot> GetSaveSlotMetaData(int index)
...
+            await UniTask.WhenAll(cloudMetaDownloadTaskList);
+        }
+
+        internal async UniTask SyncSelectedSlotData()
+        {
+            List<UniTask> datafileDownloadTaskList=new List<UniTask>();
+
+            foreach(string key in _selectedSlot.FileKeys.Keys)
+            {
+                string filePath = Path.Combine(_slotConfiguration.SavePathRoot, _selectedSlot.Name, _slotConfiguration.SlotDatafolderName,$"{key}{_slotConfiguration.SaveFileExtention}");
```
Comment (BDeshiDev) [inline Assets/Packages/com.studio23.ss2.savesystem/Runtime/Core/SaveSlotProcessor.cs:108]: Should be method on saveslot.

**Review by zubEpty — APPROVED** [review]

**Review by dipokDipu — APPROVED** [review]

**Review by N-Evan — APPROVED** [review]

**Review by BDeshiDev — APPROVED** [review]
Minor refactors for moving path construction to saveslot class recommended but not necessary atn.

**Pull request merged**